src/runtime/types.cpp

```diff
@@ -1680,6 +1681,25 @@ static Box* functionCode(Box* self, void*) {
     return codeForFunction(func);
 }
 
+static void functionSetCode(Box* self, Box* v, void*) {
+    assert(self->cls == function_cls);
+
+    if (v == NULL || !PyCode_Check(v))
+        raiseExcHelper(TypeError, "__code__ must be set to a code object");
+
+    BoxedFunction* func = static_cast<BoxedFunction*>(self);
+    BoxedCode* code = static_cast<BoxedCode*>(v);
+
+    if (!func->f->source || !code->f->source)
+        raiseExcHelper(TypeError, "__code__ can only be set on python functions");
```

Comment:
Personally I'd make this an assertion (crash) rather than an exception.  probably not that important in this case since code shouldn't be hitting this, but exceptions can get swallowed in a lot of ways :/